tox.ini

```diff
@@ -19,8 +19,16 @@ commands =
 select = F401,F402,F821,F823,F841,F811,E265,E266,F541,W605,E722,F822,F523,W291,F405
 exclude = awx/ui/node_modules,awx/ui/node_modules,env,awx_collection_build
 
+[testenv:pip-compile-docs]
+description = Compile docs build requirements
+allowlist_externals = pip-compile
+commands =
+  pip-compile --allow-unsafe --output-file=docs/docsite/requirements.txt --strip-extras docs/docsite/requirements.in
+
 [testenv:docs]
 description = Build documentation
-deps = -r{toxinidir}/docs/docsite/requirements.txt
+deps =
+  -r{toxinidir}/docs/docsite/requirements.in
+  -c{toxinidir}/docs/docsite/requirements.txt
```

Comment:
I might not know what I'm talking about, but isn't it odd to specify a requirements file _and_ a constraints file? Sure they should be consistent, but then why not just use the constraints file?

Reply:
so `-c` is for constraints file which points to `docs/docsite/requirements.txt`, but you also have `docs/docsite/constraints.in`. I understand having the `*.in` file and the materialized `*.txt` file but still trying to figure out how the 3rd file got added to the mix.

Reply:
Chances are pretty good I don't know what I'm talking about... It is kind of a tricky one but we started using pip-compile over in `ansible/ansible-documentation` and the approach seems really good to me. While I'm no expert, declaring all the constraints for direct dependencies in a separate file gives a bit more flexibility and the option to not install pinned requirements. You can see various `*.in` files in `ansible-documentation` here: https://github.com/ansible/ansible-documentation/tree/devel/tests

Maybe we don't need to go into this level of granularity with the awx docs. That's totally fine. I was hoping to get this kind of input in the review. If you think it's more straightforward we could get rid of the `constraints.in` file.

Reply:
Also, just to mention it, I would really like to upgrade the sphinx version we're using to build the docsite as a follow on to this PR. It seemed like a good idea to create a lockfile as a first step. If all goes well maybe we could also use dependabot here too.

Reply:
> Maybe we don't need to go into this level of granularity with the awx docs. That's totally fine. I was hoping to get this kind of input in the review. If you think it's more straightforward we could get rid of the `constraints.in` file.

I also noticed this and that extra bit is probably unnecessary here.

Reply:
> I might not know what I'm talking about, but isn't it odd to specify a requirements file _and_ a constraints file? Sure they should be consistent, but then why not just use the constraints file?

@AlanCoding so the constraints/lockfiles are used to restrict what the depresolver is allowed to consider. But requirements are direct user requests to install things. When you `pip install $thing`, it records that you requested it to be installed directly for future reference, it also records if some dependencies are transitive for other things. That also applies if the install requests are coming from requirements files. But constraints files don't request anything. You can add a million things there but pip will not install them, unless they are present in the dependency tree derived from the install request/requirements.

One important thing to keep in mind is that the dependency tree is different depending on the OS/Python version/CPU arch. It would only be the same if each distribution package in the entire dependency tree (including transitive deps) is pure-Python, which is rarely the case.
Some dependencies are installed depending on the Python version (backports/fallbacks usually), others are only installed on Windows and macOS but not Linux, etc.
This means that you may generate a constaints file on a different system that would contain entries that are impossible to install in other environment. And so abusing constraints as requirements (as in passing constraints to `-r`) may cause cases where `pip install` would fail.
Actually, this is the reason I like having a lot of constraint files — per each python+os+arch combo.

So to answer the confusion of having both `-r` and `-c` — it is by design that both are to be supplied.

Note that the syntax of the constraints files is almost the same as requirements files. I think the only difference is that you can specify more CLI arguments (like `--extra-index-url` or `-c`) in requirement files. I think you can "include" constraints into other constraints (IIRC) but probably not other CLI args.

Reply:
> Actually, this is the reason I like having a lot of constraint files — per each python+os+arch combo.

Here's a complex example of implementing this: https://github.com/cherrypy/cheroot/blob/00c3cc3/tox.ini#L26-L29 / https://github.com/cherrypy/cheroot/tree/00c3cc3/requirements / https://github.com/cherrypy/cheroot/tree/00c3cc3/bin / https://github.com/cherrypy/cheroot/blob/00c3cc3/.github/workflows/pip-tools.yml. But it's not easily reusable so it's probably an overkill as it is, for awx.